src/r1cs_nark_pcd/mod.rs

```diff
@@ -0,0 +1,612 @@
+use crate::error::PCDError;
+use crate::{CircuitSpecificSetupPCD, Error, PCDPredicate, PCD};
+use ark_accumulation::r1cs_nark_as::r1cs_nark::R1CSNark;
+use ark_accumulation::r1cs_nark_as::{ASForR1CSNark, AccumulatorInstance, InputInstance};
+use ark_accumulation::{
+    r1cs_nark_as, AccumulationScheme, Accumulator, AccumulatorRef, InputRef, MakeZK,
+};
+use ark_ec::CurveCycle;
+use ark_ff::PrimeField;
+use ark_r1cs_std::groups::CurveVar;
+use ark_sponge::constraints::{AbsorbableGadget, CryptographicSpongeVar};
+use ark_sponge::{Absorbable, CryptographicSponge};
+use ark_std::marker::PhantomData;
+use ark_std::rand::{CryptoRng, Rng};
+use help_circuit::HelpCircuit;
+use main_circuit::MainCircuit;
+
+/// Defines the public data structures that the [R1CSNarkPCD] uses.
+pub mod data_structures;
+use data_structures::*;
+
+/// The help circuit of the recursion.
+mod help_circuit;
+
+/// The main circuit of the recursion.
+mod main_circuit;
+
+pub(crate) const MAKE_ZK: bool = true;
+
+/// The different types needed by the [`R1CSNarkPCD`][nark_pcd] construction.
+///
+/// [nark_pcd]: crate::r1cs_nark_pcd::R1CSNarkPCD
+pub trait R1CSNarkPCDConfig<E: CurveCycle>
+where
+    E: CurveCycle,
+    MainField<E>: PrimeField + Absorbable<MainField<E>>,
+    HelpField<E>: PrimeField + Absorbable<HelpField<E>>,
+    MainAffine<E>: Absorbable<HelpField<E>>,
+    HelpAffine<E>: Absorbable<MainField<E>>,
+{
+    /// The curve var for the main affine.
+    type MainCurveVar: CurveVar<MainProjective<E>, HelpField<E>> + AbsorbableGadget<HelpField<E>>;
+
+    /// The curve var for the help affine.
+    type HelpCurveVar: CurveVar<HelpProjective<E>, MainField<E>> + AbsorbableGadget<MainField<E>>;
```

Comment:
Maybe you can extract this into a `CurveCycleGadget` trait, similar to `CurveCycle`?

Reply:
Would putting this relation into a gadget/var be appropriate? The constraint fields are different. Gadgets and vars usually specify a single constraint field right?